### ISSUE TITLE
-march=i186 or -march=any_186 do not check for 186 machine

### ISSUE THREAD

**Issue by ecm-pushbx** [tkchia/gcc-ia16#57]
Using either option to enable usage of 186+ instructions does not add code to check for an 186. The newlib-ia16 CRT startup code does have such a check but in my tests it was never compiled in: https://github.com/tkchia/newlib-ia16/blob/1c38cec61638e54f961731bf28615a91f16454d7/libgloss/ia16/dos-models-crt0.S#L37

Stepping through the CRT until it calls main also did not seem to do any such check, unless I missed it. The main function uses the `enter`, `push` immediate, and `shl` immediate instructions, that is it assumes a 186+ machine.

Here's my test case:

```
$ cat test2.c

#include <stdio.h>

int main(int argc, char** argv) {
  int volatile a = 31;
  a <<= 15;
  printf("Hello world! a=%04Xh\n", a);
  return 0;
}
$ ia16-elf-gcc test2.c -Os -o test2.exe -mcmodel=small -march=i186 && dosemu -K "$PWD" -dumb -quiet -E "c:ldebugu test2.exe"
About to Execute : c:ldebugu test2.exe
-r
AX=0000 BX=0000 CX=6660 DX=0000 SP=0000 BP=0000 SI=0000 DI=0000
DS=199B ES=199B SS=1FBE CS=19A9 IP=0020 NV UP EI PL ZR NA PE NC
19A9:0020 16                push    ss
-u
19A9:0020 16                push    ss
19A9:0021 07                pop     es
19A9:0022 31C0              xor     ax, ax
19A9:0024 BF3005            mov     di, 0530
19A9:0027 B98C00            mov     cx, 008C
19A9:002A FC                cld
19A9:002B F3AB              rep     stosw
19A9:002D 1E                push    ds
19A9:002E 16                push    ss
19A9:002F 1F                pop     ds
19A9:0030 E84A01            call    017D
19A9:0033 E84801            call    017E
19A9:0036 1F                pop     ds
19A9:0037 368C1E3805        mov     word [ss:0538], ds
19A9:003C 8E062C00          mov     es, word [002C]
-
```

**Comment by tkchia** [comment]
Hello @ecm-pushbx,

Thanks for the report.

The 186+ checking code is a bit of an undocumented and unreliable "feature".  I think I will add an explicit command line switch which a user can use to explicitly enable this check (like what I did with `-mmsdos-handle-v1`, which explicitly checks for MS-DOS v1.x).

(A bit of history:  I added the 186+ checking code back in Feb 2019, when `-march=any_186` would make `gcc-ia16` link against a whole different set of libraries and `crt0` startup files.  Later though, because there were just too many such sets of libraries, so I decided to merge some of them.  So in some cases `-march=any_186` would end up linking code against the same 8086 libraries --- as e.g. `-march=any` --- and thus not get the checking code.)

Thank you!
